### PR TITLE
Additional hooks for inventory manipulations callbacks.

### PR DESCRIPTION
**Pull request by Foghrye4**
Well, i screwed a little when i tried to remove 'minetest.clear_craft' from old master branch, so here is a new pull request.

For those, who does not saw https://github.com/minetest/minetest/pull/3962 this pull will add modding API function, which allow modders to trace any inventory changes (detached, nodemeta or player).

### CODE REVIEW

File: builtin/game/register.lua [hunk 1]
```diff
@@ -510,7 +510,15 @@ core.registered_craft_predicts, core.register_craft_predict = make_registration(
 core.registered_on_protection_violation, core.register_on_protection_violation = make_registration()
 core.registered_on_item_eats, core.register_on_item_eat = make_registration()
 core.registered_on_punchplayers, core.register_on_punchplayer = make_registration()
-
+core.registered_on_player_inventory_remove_item, core.register_on_player_inventory_remove_item = make_registration()
```
Comment (t4im) [inline builtin/game/register.lua]: Mind the conventional plural-s on *most* other callback tables.

**Comment by t4im** [comment]
If you transfer an item from your creative inventory to a chest this will trigger at least two callbacks for essentially the same action.
Having everything in 9 different callbacks likely forces people to register something multiple times.

Can't you just pass the `InvRef` as first argument? Its `get_location` gives you type (player, nodemeta, detached) and position/name. You'd have to look it up anyway, first thing in a callback.

You could even combine all 9 into a single callback:

```lua
--- passed to core.register_on_inventory_change
function(from_inv, from_list, from_slot, to_inv, to_list, to_slot, stack, player)
  maybe we can allow to interrupt the inventory action? Or maybe even change it. 
  return ... -- 
end
```

The parameters might seem a bit long, but if you account for the additional `player`, it is only 2 arguments longer than your current suggestion. And it is less verbose than having to create multiple callbacks.

Coming to think of it, why don't node callbacks do something similar?

**Comment by Foghrye4** [comment]
@t4im There is a reason for this. If you look up C++ code, you will see, that every change made passed by "InventoryList" object functions. This list have no parent inventory reference among his fields. Currently you can directly access lists and change their contents without using inventory object. To make your variant i must either add one additional field in "InventoryList" object with inventory reference or remove any possibility to access list directly. 
Making list "know" where from comes new itemstack is even more complex. Currently inventory manager just delete itemstack in an old place and re-create it on a new place. Both source and destination list have they own listeners (which could be a same) and differ this movement case from any other cases would be very difficult and bring unnecessary complexity to code.

You are free to fork my work and modify it a way you want to.

**Comment by paramat** [comment]
Please squash commits to make it easier to review.

**Comment by Foghrye4** [comment]
Pull is rewritten on fresh version and compiled silently. But I’ve got a **segmentation fault**.
Here is a **gdb** output:
http://pastebin.com/0L3Ki1JP

**Comment by sfan5** [comment]
What's the intended usage for `register_on_nodemeta_inventory_*` and `register_on_detached_inventory_*`?

**Comment by Foghrye4** [comment]
@sfan5 there is no such. Because callbacks implemented on server side inside InventoryList class, InventoryChangesReciever class, as Rubenwardy suggest, implemented in any inventory type, including detached and nodemeta. It could be used to trace all item movements for all players and mod actions. But I absolutely don't care about such movements.

**Comment by SmallJoker** [comment]
Replaced with #5647

**Pull request closed without merging**